### ISSUE TITLE
Installation and login failing in 3.6 (reverse proxy environments)

### ISSUE THREAD

**Issue by andreasotto** [k-next/kirby#3947]
_Update: the same problem occurs for login and all other panel urls, so in generally kirby 3.6 panel completely doesn't work behind reverse proxy environments any more, at least when using other port numbers internally._ 

**Describe the bug**  
Kirby can not be installed any more behind reverse proxy setups because the installation url is returned wrong.

We are generally using this `url` method in `config.php` to run kirby behind reverse proxy installations (docker environments behind traefik router):

config.php:

```
    // fix to make available behind reverse proxy
    'url' => function () {
        if(isset($_SERVER['HTTP_X_FORWARDED_FOR'])) {
            return $_SERVER['HTTP_X_FORWARDED_PROTO'] . '://' . $_SERVER['HTTP_X_FORWARDED_HOST'];
        }
        return $_SERVER['REQUEST_SCHEME'] . ':'  . '//' . $_SERVER['SERVER_NAME'] . ':' . $_SERVER['SERVER_PORT'];
    }
```
That function helps to output the correct urls using HTTP_X_FORWARDED_HOST so that kirby works behind load balancers or other kinds of reverse proxies like traefik.

The panel first time installation fails in such environments, probably because the `url` method is not used to generate the installation redirect url:

In 3.6 it seems that the configurable `url` method is not used when building the panel redirect url for the webpage. The code in the page usually looks like:

```
// Panel state
   const json = {"$config": ...... }
```

Within the json there is the panel installation like this: 
`"$url":"https:\/\/some.example.host:8080\/panel\/installation"`

The problem is **8080**:
The internal port number is used, instead of the url which is returned from the `url` method above. All other urls in the page are correctly rewritten by that method. But for this single situation (building the redirect for the first installation) the wrong url is returned.

**To Reproduce**  
1. use kirby behind traefik or another proxy
2. use the above `url` method to overwrite the urls to have kirby work in such environments
3. try to load the panel without existing accounts and expect kirby to start installation
4. See error (page stays blank, error in console log because of wrong url)

**Screenshots**  
![image](https://user-images.githubusercontent.com/2129515/142625849-afbf8985-dab6-430d-817f-41399b8131ac.png)

**Kirby Version**  
3.6

**Comment by andreasotto** [comment]
temporary workaround:

in `index.php`:
```
if(isset($_SERVER['HTTP_X_FORWARDED_FOR'])) {
  $_SERVER['SERVER_PORT'] = 443;
}
```

This helps to manually overwrite the port number as long as the request has been forwarded by a proxy (i.e. when the X-Forwarded-For is set)

(In my opinion kirby should handle the X-Forwarded-For situation internally, so that it's not necessary any more to use `url` function in `config.php` to keep kirby alive.)

**Comment by distantnative** [comment]
I would agree that we should try to fix this in `Kirby\Http\Server` and add support for the general problem. Redefining `url` component doesn't seem like it should be needed ideally.

**Comment by distantnative** [comment]
@andreasotto if you don't use this `url` config hack, how does 3.6.0 fail for you with your setup? Would be great to get that insight.

**Comment by bastianallgeier** [comment]
I had to move this into a future release, because we haven't been able to reproduce the setup yet and still need more information.

**Comment by samzzi** [comment]
> I had to move this into a future release, because we haven't been able to reproduce the setup yet and still need more information. 

I have a demo setup where I can give you ssh access if you want. Or if you tell me what you want to know I can give you more details?

**Comment by bastianallgeier** [comment]
@samzzi that access would be fantastic. Can you send me a dm on Discord?

**Comment by fabianmichael** [comment]
@bastianallgeier A blank panel can also occur in 3.6, if the request goes from e.g. `https://www.domain.tld´ to `https://domain.tld`. I had a similar problem like described above, that might be related, see: https://github.com/brocessing/kirby-webpack/issues/72

**Comment by lukasbestle** [comment]
Quick way to spin up a reverse proxy for reproduction:

1. Download the Caddy web server if you don't have it installed already. Downloading the binary from here is enough: https://github.com/caddyserver/caddy/releases
2. Spin up a reverse proxy with the following command:

```
caddy reverse-proxy --from :8123 --to kirby.test:80
```

This will spin up a reverse proxy on port `8123` that redirects traffic to port `80`.

Unfortunately this doesn't work on my machine as I use Caddy as my main development web server and something with TLS is broken. I will try to get it working tomorrow, but if you find the time you can already play around with it until then.

**Comment by ghost** [comment]
Hi, I'm also affected by this after upgrading to `3.6.1.1`. I tried to follow the recommendations in other issues too, setting the `X-Forwarded-{Proto,Port,Host,For}` to no avail.

I have the same issue as shown in the first screenshot of this issue. The one containing the `8080` port in the url for `pushState`on `History` thingy.

**Comment by jangoedde** [comment]
Hi there,

we also have the same problems with the 3.6.X version when installing in a subfolder (https://www.domain.com/path-to-kirby/) in combination with nginx.
If you need more information about reproduction, let me know.

**Comment by andreasotto** [comment]
Has there been a change in `3.6.1.1`? 

The workaround
```
if(isset($_SERVER['HTTP_X_FORWARDED_FOR'])) {
    $_SERVER['SERVER_PORT'] = 443;
}
```
doesn't work any more.
So currently i'm completely stuck.

We run kirby instances in docker containers behind traefik as a router. The docker containers with the kirby installations are running on port 8080. Traefik routes incoming tls traffic from https port `443` to local port `8080` in the docker container. The reply from kirby contains **wrongly** built urls containing the _internal_ port. 

If only I knew where kirby is miscomposing the url I would try to help fix this problem. It should really be possible to run Kirby behind a proxy.

**Comment by samzzi** [comment]
I'm pretty sure this is related so adding this for extra info on this ticket: [https://forum.getkirby.com/t/reverse-proxy-shenanigans/24722](https://forum.getkirby.com/t/reverse-proxy-shenanigans/24722)

**Comment by samzzi** [comment]
@afbora I hope you don't mind me thinking out loud here (with the risk of making a fool of myself 😅) 

I've applied your hotfix (https://github.com/getkirby/kirby/issues/3947) on the site where I'm currently having problems with and this didn't fix it. Although this really really is an edge case (reverse proxy via subfolder etc...) so less straight forward than the server I provided. This one also hasn't got different ports.

In regards of your change I was looking at the code and I'm wondering. Where is the `$forwarded` being set or passed? I see it in function syntaxes but I did a find and can't find any place where the $forwarded actually is being set or passed.

In my edge case if I parse the $forwarded it always and everywhere is false and because of that reason the HTTP_X_FORWARDED_HOST is not being used. (`host`@`kirby/src/Http/Server.php` - line 168)

If I manually (hardcoded) add `$forwarded = true;` in `kirby/src/HttpUri.php` online 245. The host does get rendered as it should. (although the panel still doesn't work in my specific case because the $url in vue isn't rendering the subfolder but this is sooo specific in this setup that I think we need to keep this out of the equation)

So I was wondering, and possibly I'm overlooking things, where does the $forwarded come from? 🤔

**Comment by samzzi** [comment]
@afbora just FYI, If I change line 199 in `kirby/src/Panel/View.php` to Panel::Url() instead of Url::current(), everything works as expected 🤔 (on my edge case) (I'm guessing because that one passes via my altered instance and Url::current() does not.

```php
/** HOTFIX reverse proxy **/
'$url' => /*Url::current()*/Panel::url(),
/** HOTFIX reverse proxy **/
```

**Comment by bastianallgeier** [comment]
Looks like we finally have a proper fix for this for 3.6.3. Sorry that it took so long. It would be fantastic if you could help us test the fix during the RC phase. The first release candidate will follow later today. 

We will re-open if it still does not work properly. But I'll keep my fingers crossed that we don't have to :)
